File: server/src/test/java/org/jetbrains/bsp/bazel/server/sync/ProjectStorageTest.java
package org.jetbrains.bsp.bazel.server.sync;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.AssertionsForClassTypes.fail;

import io.vavr.collection.HashMap;
import io.vavr.collection.HashSet;
import io.vavr.collection.List;
import io.vavr.control.Option;
import java.io.IOException;
import java.net.URI;
import java.nio.file.Files;
import org.jetbrains.bsp.bazel.logger.BspClientLogger;
import org.jetbrains.bsp.bazel.server.sync.languages.java.JavaModule;
import org.jetbrains.bsp.bazel.server.sync.languages.java.Jdk;
import org.jetbrains.bsp.bazel.server.sync.languages.scala.ScalaModule;
import org.jetbrains.bsp.bazel.server.sync.languages.scala.ScalaSdk;
import org.jetbrains.bsp.bazel.server.sync.model.Label;
import org.jetbrains.bsp.bazel.server.sync.model.Language;
import org.jetbrains.bsp.bazel.server.sync.model.Module;
import org.jetbrains.bsp.bazel.server.sync.model.Project;
import org.jetbrains.bsp.bazel.server.sync.model.SourceSet;
import org.jetbrains.bsp.bazel.server.sync.model.Tag;
import org.jetbrains.bsp.bazel.utils.dope.DopeTemp;
import org.junit.jupiter.api.Test;

public class ProjectStorageTest {

  @Test
  public void shouldStoreAndLoadProject() throws IOException {
    var path = DopeTemp.INSTANCE.createTempPath("project-cache-test.json");
    Files.deleteIfExists(path);
    var storage = new FileProjectStorage(path, new BspClientLogger());

    var empty = storage.load();
    assertThat(empty).isEmpty();

    var project =
        new Project(
            URI.create("file:///root"),
            List.of(
                new Module(
                    Label.from("//project:project"),
                    false,
                    List.of(Label.from("//project:dep")),
                    HashSet.of(Language.JAVA),
                    HashSet.of(Tag.LIBRARY),
                    URI.create("file:///root/project"),
                    new SourceSet(
                        HashSet.of(URI.create("file:///root/project/Lib.java")),
                        HashSet.of(URI.create("file:///root/project/"))),
                    HashSet.empty(),
                    HashSet.empty(),
                    Option.some(
                        new ScalaModule(
                            new ScalaSdk("org.scala", "2.12.3", "2.12", List.empty()),
                            List.empty(),
                            Option.some(
                                new JavaModule(
                                    new Jdk("8", Option.none()),
                                    Option.none(),
                                    List.empty(),
                                    List.empty(),
                                    URI.create("file:///tmp/out"),
                                    List.empty(),
                                    Option.none(),
                                    List.empty(),
                                    List.empty(),
                                    List.empty(),
                                    List.empty(),
                                    List.empty())))))),
            HashMap.of(URI.create("file:///root/project/Lib.java"), Label.from("file:///root")));

    storage.store(project);
    var loaded = storage.load();
    loaded.forEach(actual -> assertThat(actual).isEqualTo(project));
    loaded.onEmpty(() -> fail("Project not loaded"));
  }
}

File: server/src/test/kotlin/org/jetbrains/bsp/bazel/server/sync/ProjectStorageTest.kt
package org.jetbrains.bsp.bazel.server.sync

import ch.epfl.scala.bsp4j.*
import io.kotest.matchers.shouldBe
import org.jetbrains.bsp.bazel.bazelrunner.BazelRelease
import org.jetbrains.bsp.bazel.logger.BspClientLogger
import org.jetbrains.bsp.bazel.server.sync.languages.java.JavaModule
import org.jetbrains.bsp.bazel.server.sync.languages.java.Jdk
import org.jetbrains.bsp.bazel.server.sync.languages.scala.ScalaModule
import org.jetbrains.bsp.bazel.server.sync.languages.scala.ScalaSdk
import org.jetbrains.bsp.bazel.server.sync.model.Label
import org.jetbrains.bsp.bazel.server.sync.model.Language
import org.jetbrains.bsp.bazel.server.sync.model.Module
import org.jetbrains.bsp.bazel.server.sync.model.Project
import org.jetbrains.bsp.bazel.server.sync.model.SourceSet
import org.jetbrains.bsp.bazel.server.sync.model.Tag
import org.junit.jupiter.api.Test
import java.io.File
import java.io.IOException
import java.net.URI

class ProjectStorageTest {
    @Test
    @Throws(IOException::class)
    fun `should store and load project`() {
        val file = File.createTempFile("project-cache-test", ".json")
        file.delete()
        file.deleteOnExit()
        val path = file.toPath()
        val mockBuildClient = object : BuildClient {
            override fun onBuildShowMessage(p0: ShowMessageParams?) {
            }

            override fun onBuildLogMessage(p0: LogMessageParams?) {
            }

            override fun onBuildPublishDiagnostics(p0: PublishDiagnosticsParams?) {
            }

            override fun onBuildTargetDidChange(p0: DidChangeBuildTarget?) {
            }

            override fun onBuildTaskStart(p0: TaskStartParams?) {
            }

            override fun onBuildTaskProgress(p0: TaskProgressParams?) {
            }

            override fun onBuildTaskFinish(p0: TaskFinishParams?) {
            }

            override fun onRunPrintStdout(p0: PrintParams?) {
            }

            override fun onRunPrintStderr(p0: PrintParams?) {
            }

        }
        val storage = FileProjectStorage(path, BspClientLogger(mockBuildClient))
        val empty = storage.load()
        empty shouldBe null
        val scalaModule = ScalaModule(
            ScalaSdk("org.scala", "2.12.3", "2.12", emptyList()), emptyList(), JavaModule(
                Jdk("8", null),
                null,
                emptyList(),
                emptyList(),
                URI.create("file:///tmp/out"),
                emptyList(),
                null,
                emptyList(),
            )
        )
        val project = Project(
            URI.create("file:///root"),
            listOf(
                Module(
                    Label("//project:project"),
                    false,
                    listOf(Label("//project:dep")),
                    hashSetOf(Language.JAVA),
                    hashSetOf(Tag.LIBRARY),
                    URI.create("file:///root/project"),
                    SourceSet(
                        hashSetOf(URI.create("file:///root/project/Lib.java")),
                        emptySet(),
                        hashSetOf(URI.create("file:///root/project/"))
                    ),
                    emptySet(),
                    emptySet(),
                    emptySet(),
                    scalaModule,
                    hashMapOf()
                )
            ),
            mapOf(URI.create("file:///root/project/Lib.java") to Label("file:///root")),
            emptyMap(),
            emptyList(),
            BazelRelease(7),
        )
        storage.store(project)
        val loaded = storage.load()
        loaded shouldBe project
    }
}
